### PR TITLE
feat(testkit): allows for unit testing when cross-component dependencies exist

### PR DESCRIPTION
**Pull request by efgpinto**
References #984 

This is still in a draft version, opening for visibility only (there might be refactors still left to do as I tend to do those at the end).

Usage example with `javasdk`:
```java
    when(shoppingCartService.addItem(any()))
            .thenReturn(CompletableFuture.completedFuture(Empty.getDefaultInstance()));
    var mockRegistry = TestKitMockRegistry.withMock(ShoppingCartService.class, shoppingCartService);
    var service = ShoppingCartActionImplTestKit.of(ShoppingCartActionImpl::new, mockRegistry);
```

Usage example with `scalasdk`:
```scala
      val mockCounter = mock[CounterService]
      (mockCounter.increase _)
        .expects(*)
        .returning(Future.successful(Empty()))
      val mockRegistry = TestKitMockRegistry.of(Set(mockCounter))

      val service = ExternalCounterActionTestKit(new ExternalCounterAction(_), mockRegistry)
```

Tasks:

- [x] Scala test kit gen to include mock registry
- [x] Java test kit gen to include mock registry
- [x] Scala sample project with unit test with cross-component dependency
- [x] Java sample project with unit test with cross-component dependency
- [x] Scala example with cross-service call
- [x] overall review and refactor

### CODE REVIEW

File: sdk/java-sdk-testkit/src/main/scala/kalix/javasdk/testkit/impl/TestKitMockRegistry.scala
Comment (johanandren) [inline sdk/java-sdk-testkit/src/main/scala/kalix/javasdk/testkit/impl/TestKitMockRegistry.scala:32]: This is probably internal, and leaks `Option` to Java, in many other components we use a interface/factories in the public API and keep methods needed for the internals in the implementation that a user does not really access. See for example `kalix.javasdk.Metadata`

File: sdk/scala-sdk-testkit/src/main/scala/kalix/scalasdk/testkit/impl/TestKitMockRegistryImpl.scala
Comment (efgpinto) [inline sdk/scala-sdk-testkit/src/main/scala/kalix/scalasdk/testkit/impl/TestKitMockRegistryImpl.scala:30]: Notice how on the `scalasdk` I'm using a `Set` and not a `Map` (used in the `javasdk)`... The end goal would be to choose between one or the other version and make it consistent on both sdks ... I've kept both options for the sake of the review because I'm not quite sure. Although this version with the `Set` allows the user to do only `TestKitMockRegistry.withMock(mockA)` instead of `TestKitMockRegistry.withMock(A.class, mockA)` I wonder if this convoluted logic is a bit too risky. @johanandren @octonato

File: sdk/scala-sdk-testkit/src/main/scala/kalix/scalasdk/testkit/TestKitMockRegistry.scala
Comment (efgpinto) [inline sdk/scala-sdk-testkit/src/main/scala/kalix/scalasdk/testkit/TestKitMockRegistry.scala:25]: I wonder if we should call this just `MockRegistry` instead?

**Review by efgpinto — COMMENTED** [review]
Thanks @johanandren for the valuable feedback, I have now addressed your suggestions but please feel free to call out any improvement you might spot.

3 points remaining:
1.  I would like some input in choosing between using a Map or a Set on the MockRegistryImpl (I've left a comment in-place to discuss that). You can compare the implementations for javasdk (using a `Map`) and the scalasdk (using a `Set`)
2. still need to figure out why the javadocs gen is failing on CI
3. it could be useful to provide an example on how to take advantage of this when generating the testkit unit examples but not sure that should be done in this PR or later when we do proper documentation (@octonato mentioned the docs would be done in another issue)

**Review by johanandren — COMMENTED** [review]
Lookin' good, some more feedback added!

**Review by johanandren — APPROVED** [review]
LGTM

**Review by octonato — COMMENTED** [review]
Looking good @efgpinto

I left some suggestions. 

@johanandren, it will be good to have your input on my suggestions as well.

File: sdk/java-sdk-testkit/src/main/scala/kalix/javasdk/testkit/impl/TestKitEventSourcedEntityContext.scala
Comment (johanandren) [reply sdk/java-sdk-testkit/src/main/scala/kalix/javasdk/testkit/impl/TestKitEventSourcedEntityContext.scala:26]: Let's open an issue and discuss that separately.

Comment (octonato) [reply sdk/java-sdk-testkit/src/main/scala/kalix/javasdk/testkit/impl/TestKitEventSourcedEntityContext.scala:26]: sure, that was not the intention to add this to this PR.

File: sdk/java-sdk-testkit/src/main/java/kalix/javasdk/testkit/MockRegistry.java
Comment (efgpinto) [reply sdk/java-sdk-testkit/src/main/java/kalix/javasdk/testkit/MockRegistry.java:57]: For this java one, I believe we will have to go with something like but LGTM if you agree
```java
  MockRegistry.create()
    .withMock(...)
    .withMock(...)
```

File: sdk/java-sdk-testkit/src/main/scala/kalix/javasdk/testkit/impl/TestKitEventSourcedEntityContext.scala
Comment (octonato) [reply sdk/java-sdk-testkit/src/main/scala/kalix/javasdk/testkit/impl/TestKitEventSourcedEntityContext.scala:26]: Yeah, I think for now we need to go with it. If we decide to deprecate, we still need to keep it around for a while and usage in the entity testkit will still be valid.

**Review by octonato — APPROVED** [review]
LGTM

I will merge it. We can with documentation on follow-up PRs.

**Pull request merged**
